File: lecture_summarizer/templates.py
from __future__ import annotations
from pathlib import Path
from functools import lru_cache
from textwrap import dedent

from .constants import TEMPLATES_DIR


DEFAULT_PREAMBLES = {
    "slides": dedent(
        r"""\documentclass[aspectratio=43]{beamer}

\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{amsfonts}
\usepackage{tikz}
\usepackage{xcolor}
\usepackage{graphicx}
\usepackage{hyperref}

\usetheme{Madrid}
\setbeamertemplate{navigation symbols}{}

\title{}
\author{}
\date{}

\begin{document}
"""
    ).strip()
    + "\n",
    "lecture": dedent(
        r"""\documentclass{article}

\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{amsfonts}
\usepackage{amsthm}
\usepackage{physics}
\usepackage{bm}
\usepackage{geometry}
\geometry{margin=1in}

\title{}
\author{}
\date{}

\begin{document}
"""
    ).strip()
    + "\n",
    "document": dedent(
        r"""\documentclass{article}

\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{amsfonts}
\usepackage{amsthm}
\usepackage{graphicx}
\usepackage{tabularx}
\usepackage{booktabs}
\usepackage{xcolor}
\usepackage{enumitem}

\title{}
\author{}
\date{}

\begin{document}
"""
    ).strip()
    + "\n",
    "image": dedent(
        r"""\documentclass{article}

\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{amsfonts}
\usepackage{amsthm}
\usepackage{geometry}
\geometry{margin=1in}

\begin{document}
"""
    ).strip()
    + "\n",
    "video": dedent(
        r"""\documentclass{article}

\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{amsfonts}
\usepackage{amsthm}
\usepackage{xcolor}
\usepackage{enumitem}
\usepackage{geometry}
\geometry{margin=1in}

\begin{document}
"""
    ).strip()
    + "\n",
}


class TemplateLoader:
    def __init__(self, base: Path | None = None):
        self.base = Path(base or TEMPLATES_DIR)

    def _load_optional(self, name: str) -> str | None:
        path = self.base / name
        if not path.exists():
            return None
        return path.read_text()

    def _load_or_default(self, name: str, default: str | None) -> str:
        text = self._load_optional(name)
        if text is not None:
            return text
        if default is None:
            raise FileNotFoundError(f"Template not found: {self.base / name}")
        return default

    @lru_cache(maxsize=None)
    def slide_preamble(self) -> str:
        return self._load_or_default("slide-template.txt", DEFAULT_PREAMBLES["slides"])

    @lru_cache(maxsize=None)
    def lecture_preamble(self) -> str:
        return self._load_or_default("lecture-template.txt", DEFAULT_PREAMBLES["lecture"])

    @lru_cache(maxsize=None)
    def document_preamble(self) -> str:
        return self._load_or_default("document-template.txt", DEFAULT_PREAMBLES["document"])

    @lru_cache(maxsize=None)
    def image_preamble(self) -> str:
        return self._load_or_default("image-template.txt", DEFAULT_PREAMBLES["image"])

    @lru_cache(maxsize=None)
    def video_preamble(self) -> str:
        return self._load_or_default("video-template.txt", DEFAULT_PREAMBLES["video"])

    @lru_cache(maxsize=None)
    def latex_to_md_prompt(self) -> str:
        default = dedent(
            """Convert the LaTeX source into Markdown while preserving structure.
- Keep headings mapping section -> #, subsection -> ##.
- Preserve math using $...$ or $$...$$.
- Use bullet/numbered lists for itemize/enumerate.
- Render tables as GitHub-flavored Markdown tables where possible.
- Replace images or TikZ drawings with `[Placeholder: description]`.
- Remove LaTeX-only preamble commands.

Return only the Markdown.
"""
        ).strip()
        return self._load_or_default("latex-to-md-template.txt", default)

    @lru_cache(maxsize=None)
    def latex_to_json_prompt(self) -> str:
        default = dedent(
            """Convert the LaTeX table or structured content into well-formed JSON.
- Use the first row as headers when available.
- Preserve numeric types where obvious, otherwise use strings.
- Output a JSON array of objects.
- Do not include explanations.
"""
        ).strip()
        return self._load_or_default("latex-to-json-template.txt", default)

    @lru_cache(maxsize=None)
    def prompt(self, name: str, *, default: str) -> str:
        fname = f"{name}-prompt.txt"
        if (text := self._load_optional(fname)) is not None:
            return text
        return default


from .prompts.default import DEFAULT_PROMPTS as _DEFAULT_PROMPTS

# Backwards compatibility: expose default prompts keyed by template name.
DEFAULT_PROMPTS = {kind.value: prompt for kind, prompt in _DEFAULT_PROMPTS.items()}

File: src/templates.rs
use std::collections::HashMap;
use std::fs;
use std::path::{Path, PathBuf};
use std::sync::{Arc, Mutex};

use crate::core::Kind;

#[derive(Debug, Clone)]
pub struct TemplateLoader {
    base: Arc<PathBuf>,
    cache: Arc<Mutex<HashMap<String, String>>>,
}

impl TemplateLoader {
    pub fn new(base: impl Into<PathBuf>) -> Self {
        Self {
            base: Arc::new(base.into()),
            cache: Arc::new(Mutex::new(HashMap::new())),
        }
    }

    fn load_cached(
        &self,
        key: &str,
        loader: impl FnOnce(&Path) -> Option<String>,
    ) -> Option<String> {
        if let Some(value) = self.cache.lock().unwrap().get(key).cloned() {
            return Some(value);
        }
        let result = loader(&self.base);
        if let Some(ref text) = result {
            self.cache
                .lock()
                .unwrap()
                .insert(key.to_string(), text.clone());
        }
        result
    }

    fn load_or_default(&self, filename: &str, default: &str) -> String {
        let key = format!("template::{filename}");
        if let Some(value) = self.load_cached(&key, |base| read_file(base.join(filename))) {
            return value;
        }
        default.to_string()
    }

    pub fn slide_preamble(&self) -> String {
        self.load_or_default("slide-template.txt", DEFAULT_PREAMBLES.slides)
    }

    pub fn lecture_preamble(&self) -> String {
        self.load_or_default("lecture-template.txt", DEFAULT_PREAMBLES.lecture)
    }

    pub fn document_preamble(&self) -> String {
        self.load_or_default("document-template.txt", DEFAULT_PREAMBLES.document)
    }

    pub fn image_preamble(&self) -> String {
        self.load_or_default("image-template.txt", DEFAULT_PREAMBLES.image)
    }

    pub fn video_preamble(&self) -> String {
        self.load_or_default("video-template.txt", DEFAULT_PREAMBLES.video)
    }

    pub fn latex_to_md_prompt(&self) -> String {
        self.load_or_default("latex-to-md-template.txt", DEFAULT_CONVERSIONS.latex_to_md)
    }

    pub fn latex_to_json_prompt(&self) -> String {
        self.load_or_default(
            "latex-to-json-template.txt",
            DEFAULT_CONVERSIONS.latex_to_json,
        )
    }

    pub fn markdown_to_json_prompt(&self) -> String {
        self.load_or_default(
            "markdown-to-json-template.txt",
            DEFAULT_CONVERSIONS.markdown_to_json,
        )
    }

    pub fn prompt(&self, kind: Kind, default: &str) -> String {
        let filename = format!("{}-prompt.txt", kind.as_str());
        self.load_or_default(&filename, default)
    }
}

fn read_file(path: PathBuf) -> Option<String> {
    match fs::read_to_string(path) {
        Ok(text) => Some(text),
        Err(_) => None,
    }
}

struct DefaultPreambles {
    slides: &'static str,
    lecture: &'static str,
    document: &'static str,
    image: &'static str,
    video: &'static str,
}

struct DefaultConversions {
    latex_to_md: &'static str,
    latex_to_json: &'static str,
    markdown_to_json: &'static str,
}

const SLIDES_PREAMBLE: &str = r"# Slide Deck Summary

";

const LECTURE_PREAMBLE: &str = r"# Lecture Summary

";

const DOCUMENT_PREAMBLE: &str = r"# Document Summary

";

const IMAGE_PREAMBLE: &str = r"# Image Analysis

";

const VIDEO_PREAMBLE: &str = r"# Video Summary

";

const LATEX_TO_MD_PROMPT: &str = r"Convert the LaTeX source into Markdown while preserving structure.
- Keep headings mapping section -> #, subsection -> ##.
- Preserve math using $...$ or $$...$$.
- Use bullet/numbered lists for itemize/enumerate.
- Render tables as GitHub-flavored Markdown tables where possible.
- Replace images or TikZ drawings with `[Placeholder: description]`.
- Remove LaTeX-only preamble commands.

Return only the Markdown.
";

const LATEX_TO_JSON_PROMPT: &str = r"Convert the LaTeX table or structured content into well-formed JSON.
- Use the first row as headers when available.
- Preserve numeric types where obvious, otherwise use strings.
- Output a JSON array of objects.
- Do not include explanations.
";

const MARKDOWN_TO_JSON_PROMPT: &str = r"Convert the Markdown tables or structured lists into well-formed JSON.
- Use the first row of each table as headers when available.
- Preserve numeric types where obvious, otherwise use strings.
- Output a JSON array of objects.
- Ignore narrative sections that do not map cleanly to data rows.
- Do not include explanations.
";

static DEFAULT_PREAMBLES: DefaultPreambles = DefaultPreambles {
    slides: SLIDES_PREAMBLE,
    lecture: LECTURE_PREAMBLE,
    document: DOCUMENT_PREAMBLE,
    image: IMAGE_PREAMBLE,
    video: VIDEO_PREAMBLE,
};

static DEFAULT_CONVERSIONS: DefaultConversions = DefaultConversions {
    latex_to_md: LATEX_TO_MD_PROMPT,
    latex_to_json: LATEX_TO_JSON_PROMPT,
    markdown_to_json: MARKDOWN_TO_JSON_PROMPT,
};
